### DISCUSSION TITLE
500 error when FB crawler attempts to fetch statically generated page

### DISCUSSION THREAD

**Discussion by Elxx** [vercel/vercel#4423] [Help] [answered]
Apologies if this is the wrong project to post on, as I'm not sure whether this is a Next.js or Vercel issue although I'm leaning towards Vercel.

I'm currently trying to run a Next.js site deployed to Vercel through the [Facebook sharing debugger](https://developers.facebook.com/tools/debug/?q=https%3A%2F%2Ftest-enoent-error-mv26isp7z.now.sh%2F) to ensure that my opengraph tags are correct, but every page I try returns a 500 error. In case the FB dev site requires a login, here's a [gist of the response](https://gist.github.com/Elxx/5487fc319d29145afc0be0f429930c07) that the Facebook crawler receives when fetching https://test-enoent-error-mv26isp7z.now.sh/ .

Looking at the Serverless Function log when the FB crawler hits the home page, this error pops up:
`[GET] /
20:32:12:01
2020-05-19T18:32:12.061Z	ab868153-4f87-41e6-a210-05137883309f	ERROR	Error: ENOENT: no such file or directory, scandir '/var/task/_posts'
    at Object.readdirSync (fs.js:872:3)
    at getPostSlugs (/var/task/.next/serverless/pages/index.js:3315:52)
    at getAllPosts (/var/task/.next/serverless/pages/index.js:3343:17)
    at getStaticProps (/var/task/.next/serverless/pages/index.js:823:54)
    at renderToHTML (/var/task/node_modules/next/dist/next-server/server/render.js:277:30)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async renderReqToHTML (/var/task/.next/serverless/pages/index.js:980:22)
    at async Server.render (/var/task/.next/serverless/pages/index.js:1027:22) {
  errno: -2,
  syscall: 'scandir',
  path: '/var/task/_posts'
}`

The site works perfectly fine in any web browser and doesn't throw any errors when browsed regularly.

This issue can be reproduced with a totally unmodified copy of the [next.js blog-starter](https://github.com/zeit/next.js/tree/canary/examples/blog-starter) repo deployed to Vercel with default settings. I've created an example deployment here:
https://test-enoent-error-mv26isp7z.now.sh/

This brings up a few questions:
1) Why is Vercel (or Next.js?) running a function to server-side-render the page at request time? pages/index.js uses getStaticProps and doesn't have any dynamic routes, shouldn't it just generate a .html file at build time and never touch it again?
2) But if this is intended behavior after all, I was under the impression that using node's fs module is fine in getStaticProps (since the example project does it), but the error suggests that files are inaccessible. Is this not supported in the Vercel environment?
3) What is it about the FB crawler that specifically causes this behavior? Is it skipping the CDN? I've tried emulating the same behavior in my browser with no luck, but maybe someone on the Vercel team has access to better logging to diagnose what's happening.

**Comment by styfle** [comment]
Thanks for the repro steps, we are investigating!

This kind of a "perfect storm" where facebook somehow trigged [Preview Mode](https://nextjs.org/docs/advanced-features/preview-mode) fallback, likely with an additional header.

The 500 is probably related to a bug with `process.cwd()` being evaluated incorrectly in the function when compared to the static build.

We'll get this fixed, thanks!

**Reply by 4vanger** [reply]
I'm having same bug - `process.cwd()` points to `/var/task` in deployment.
Code:
```
export const getStaticProps: GetStaticProps<AgentPageProps> = async ({ params }) => {
  const dataDirectory = join(process.cwd(), 'public')
  const fullPath = join(dataDirectory, 'agents.json')
  const readFile = util.promisify(fs.readFile)
  const json = JSON.parse(await readFile(fullPath, 'utf8'))
}
```
Getting error in runtime: 
```
2020-05-22T22:48:29.105Z	aaa0635b-3579-495c-81f4-dbf590777042	ERROR	[Error: ENOENT: no such file or directory, open '/var/task/public/agents.json'] {
  errno: -2,
  syscall: 'open',
  path: '/var/task/public/agents.json'
}
```

**Reply by styfle** [reply]
@Elxx The fallback issue with the `Range` header has been resolved. You should be able to use the blog-starter example with the Facebook Sharing Debugger. Thanks!

**Reply by akobor-bitrise** [reply]
Hi @styfle,
I'm still getting an error, but now it looks more like a timeout :( 
https://developers.facebook.com/tools/debug/?q=https%3A%2F%2Fakobor.me%2Fposts%2Fwhy-kotlin-for-backend-is-a-real-thing
Additionally, there isn't any error in function log for the given period

**Reply by goenning** [reply]
Same here. I'm getting a 206 response back now (instead of 500), but Facebook errors out with `The request to scrape the URL timed out.`

**Comment by YIPG** [comment]
I have the same issue.

code `pages/api/word.ts`
```javascript
const dictionaryPath = path.join(process.cwd(), "assets/dictionary.json")

const fileName = fs.readFileSync(dictionaryPath, "utf-8")
```

|dev|local build|Vercel|
|-|-|-|
| 👍 | 👍 | 💣 |

Getting Error (500)

```
[POST] /api/word
10:42:31:01
Status:
500
Duration:
23.64ms
Memory Used:
69 MB
ID:
9rh7g-1590370951012-8988a8755c18
User Agent:
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_4) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/81.0.4044.138 Safari/537.36
2020-05-25T01:42:31.099Z	1288b528-e70b-4d0b-afff-8700ef6f6daa	ERROR	Error: ENOENT: no such file or directory, open '/var/task/assets/dictionary.json'
    at Object.openSync (fs.js:458:3)
    at Object.readFileSync (fs.js:360:35)
    at Module.PicC (/var/task/.next/serverless/pages/api/word.js:349:40)
    at __webpack_require__ (/var/task/.next/serverless/pages/api/word.js:23:31)
    at Server.module.exports.Mr1m.__webpack_exports__.default (/var/task/.next/serverless/pages/api/word.js:209:28) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/var/task/assets/dictionary.json'
}
```

**Comment by eddyw** [comment]
@styfle today a client reported an issue with Facebook not showing thumbnails (open graph) anymore when sharing links.

For us, Facebook scrapper debugger shows these warnings:
<img width="1000" alt="Screen Shot 2020-06-11 at 12 45 07 AM" src="https://user-images.githubusercontent.com/1407526/84295152-cead9c00-ab7c-11ea-8ed3-cfd261f0e5b0.png">
Instead of 500, it's timeout now. Even though we actually set those tags (open graph meta tags).

The scrapper should show the raw HTML but it doesn't (anymore) so Facebook cannot read anything:
https://developers.facebook.com/tools/debug/echo/?q=https%3A%2F%2Fwww.productsup.com%2Fblog%2Fevaluate-feed-management-software%2F

Interestingly enough, this is the same case for nextjs.org pages:
https://developers.facebook.com/tools/debug/echo/?q=https%3A%2F%2Fnextjs.org%2Fdocs%2Fapi-reference%2Fnext.config.js%2FexportPathMap (empty)

https://developers.facebook.com/tools/debug/?q=https%3A%2F%2Fnextjs.org%2Fdocs%2Fapi-reference%2Fnext.config.js%2FexportPathMap (see warnings here)

The only reason I think it shows thumbnails when sharing this link is because it was cached. For us, only previously shared links continue to display thumbnails (open graph). Then, all vercel hosted pages timeout.

**Comment by styfle** [comment, marked as answer]
Thanks for letting us know!
The timeout has been fixed 👍

**Reply by adamkobor** [reply]
Thanks for the quick fix, you are amazing 💯

**Comment by Elxx** [comment]
Everything works beautifully now, thanks for taking the time to look into this and fix it! :)
